pyccel/codegen/codegen.py

```diff
@@ -9,33 +9,51 @@
 from pyccel.codegen.printing.fcode  import FCodePrinter
 from pyccel.codegen.printing.ccode  import CCodePrinter
 from pyccel.codegen.printing.pycode import PythonCodePrinter
+from pyccel.codegen.printing.cucode import CudaCodePrinter
 
 from pyccel.ast.core      import FunctionDef, Interface, ModuleHeader
 from pyccel.errors.errors import Errors
 from pyccel.utilities.stage import PyccelStage
 
-_extension_registry = {'fortran': 'f90', 'c':'c',  'python':'py'}
-_header_extension_registry = {'fortran': None, 'c':'h',  'python':None}
+_extension_registry = {'fortran': 'f90', 'c':'c',  'python':'py', 'cuda':'cu'}
+_header_extension_registry = {'fortran': None, 'c':'h',  'python':None, 'cuda':'h'}
 printer_registry    = {
                         'fortran':FCodePrinter,
                         'c':CCodePrinter,
-                        'python':PythonCodePrinter
+                        'python':PythonCodePrinter,
+                        'cuda':CudaCodePrinter
                       }
 
 pyccel_stage = PyccelStage()
 
 class Codegen(object):
 
-    """Abstract class for code generator."""
+    """
+    Class which handles the generation of code.
 
-    def __init__(self, parser, name):
-        """Constructor for Codegen.
+    The class which handles the generation of code. This is done by creating an appropriate class
+    inheriting from `CodePrinter` and using it to create strings describing the code that should
+    be printed. This class then takes care of creating the necessary files.
 
-        parser: pyccel parser
+    Parameters
+    ----------
+    parser : BasicParser
+        The Pyccel parser node.
+    name : str
+        Name of the generated module or program.
+    """
 
+    def __init__(self, parser, name):
+        """
+        Constructor for Codegen.
 
+        Parameters
+        ----------
+        parser: pyccel parser
```

Comment:
`pyccel parser` is not a type. We don't usually bother with docstrings for `__init__`. They are not typically viewed by users using the `?` functionality in ipython and they are always next to the similar class docstring in the code

Reply:
ah didn't notice it was `pyccel parser` as i only wrote the description under it without giving it much thought.
I have a question if i want to give the parser a type it should be `BasicParser` right ?
removed in [db21fa9](https://github.com/pyccel/pyccel-cuda/pull/32/commits/db21fa919b4b3a3defa1dc7f81325baf229694e0).

Reply:
It depends on the context. If you want a syntactic parser then you can use `SyntacticParser`, if you want a semantic parser then you can use `SemanticParser`, if you are happy with either then it should be `BasicParser`. There is also the class `pyccel.parser.Parser` which holds an instance of `SyntacticParser` and an instance of `SemanticParser`. This is used in the pipeline.

Here I would expect it to be `SemanticParser`. I don't think we can print using only the results of the `SyntacticParser` so `BasicParser` seems too general to me

Reply:
[e7c8748](https://github.com/pyccel/pyccel-cuda/pull/32/commits/e7c8748b0813137551f57469dab539cfa0971ef9)